File: java-home-inventory-core/src/br/com/sailboat/homeinventory/core/interactor/product/SaveProduct.java
package br.com.sailboat.homeinventory.core.interactor.product;

import br.com.sailboat.homeinventory.core.entity.Product;
import br.com.sailboat.homeinventory.core.interactor.UseCase;
import br.com.sailboat.homeinventory.core.repository.ProductRepository;

public class SaveProduct implements UseCase<Boolean> {

    private Product product;
    private ProductRepository productRepository;

    public SaveProduct(Product product, ProductRepository productRepository) {
        this.product = product;
        this.productRepository = productRepository;
    }

    @Override
    public Boolean execute() throws Exception {
        new ProductValidator(product).execute();
        productRepository.save(product);
        return true;
    }

}

File: android-home-inventory-app/app/src/main/kotlin/br/com/sailboat/homeinventory/domain/usecase/product/SaveProduct.kt
package br.com.sailboat.homeinventory.domain.usecase.product

import br.com.sailboat.homeinventory.domain.entity.EntityHelper
import br.com.sailboat.homeinventory.domain.entity.Product
import br.com.sailboat.homeinventory.domain.repository.ProductRepository
import br.com.sailboat.homeinventory.domain.usecase.UseCase
import javax.inject.Inject

class SaveProduct @Inject constructor(private val productRepository: ProductRepository) : UseCase<Product, Unit>() {

    override fun execute(product: Product) {
        if (product.id == EntityHelper.NO_ID) {
            productRepository.insert(product)
        } else {
            productRepository.update(product)
        }
    }

}